File: src/main/java/dev/koen/aoc/day2/One.java
package dev.koen.aoc.day2;

import dev.koen.aoc.util.FileReader;
import dev.koen.aoc.util.Range;

import java.nio.file.Path;

public class One {

    public static void main(String[] args) {
        final var count = FileReader.readLines(Path.of("src/test/resources/input-two.txt"))
                .stream()
                .filter(One::hasValidPassword)
                .count();

        System.out.println(count);
    }

    private static boolean hasValidPassword(String line) {
        final var parts = line.split(" ");
        final var range = readRange(parts[0]);
        final var c = parts[1].charAt(0);

        final var count = parts[2].chars()
                .filter(current -> current == c)
                .count();

        return range.contains(count);
    }

    private static Range readRange(String part) {
        final var numbers = part.split("-");
        final var min = Integer.parseUnsignedInt(numbers[0]);
        final var max = Integer.parseUnsignedInt(numbers[1]);
        return new Range(min, max);
    }
}

File: src/main/kotlin/dev/koen/aoc/day2/One.kt
package dev.koen.aoc.day2

import java.io.File

fun main() {
    File("src/test/resources/input-two.txt")
        .readLines()
        .map(String::toPolicy1)
        .filter(Policy1::isValid)
        .count()
        .let { amountOfValidPasswords -> println(amountOfValidPasswords) }
}

private fun String.toPolicy1(): Policy1 {
    val parts = this.split(" ")

    val range = parts[0].toIntRange()
    val c = parts[1].toCharArray()[0]
    val password = parts[2]

    return Policy1(range, c, password)
}

private fun String.toIntRange(): IntRange {
    val numbers = this.split("-")

    val start = numbers[0].toInt()
    val end = numbers[1].toInt()

    return IntRange(start, end)
}

private data class Policy1(val range: IntRange, val char: Char, val password: String) {
    fun isValid(): Boolean {
        return password.toCharArray()
            .filter(char::equals)
            .count()
            .let { range.contains(it) }
    }
}